File: backend/java/cretas-api/src/main/java/com/cretas/aims/entity/Attachment.java
package com.cretas.aims.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import org.hibernate.annotations.Where;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * 通用附件实体 (多态: entity_type + entity_id)
 *
 * <p>支持 5 大业务场景的附件挂载: 客户跟踪 / 采购订单 / 质检 / 生产证据 / 财务凭证. 决策见
 * {@code 宏见竞品分析/01-客户档案/SCHEMA_DESIGN.md §2.3}: 多态关联不强外键, 文件实体仅存
 * OSS URL (不存 BLOB), 权限跟随 entity, 下载通过 OSS pre-signed URL 实现细粒度控制.
 *
 * @author Cretas Team — Track C
 * @since 2026-05-15 (C-ATT-1)
 */
@Data
@EqualsAndHashCode(callSuper = true)
@SuperBuilder
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "attachments", indexes = {
        @Index(name = "idx_att_entity", columnList = "factory_id,entity_type,entity_id"),
        @Index(name = "idx_att_uploader", columnList = "factory_id,uploaded_by"),
        @Index(name = "idx_att_tag", columnList = "factory_id,business_tag"),
        @Index(name = "idx_att_hash", columnList = "factory_id,file_hash")
})
@Where(clause = "deleted_at IS NULL")
public class Attachment extends BaseEntity {

    @Id
    @Column(name = "id", nullable = false, length = 191)
    private String id;

    @PrePersist
    void assignUUID() {
        if (id == null) {
            id = UUID.randomUUID().toString();
        }
        if (uploadedAt == null) {
            uploadedAt = LocalDateTime.now();
        }
    }

    @Column(name = "factory_id", nullable = false, length = 50)
    private String factoryId;

    @Enumerated(EnumType.STRING)
    @Column(name = "entity_type", nullable = false, length = 50)
    private EntityType entityType;

    @Column(name = "entity_id", nullable = false, length = 191)
    private String entityId;

    @Column(name = "file_name", nullable = false, length = 255)
    private String fileName;

    @Column(name = "file_url", nullable = false, length = 1000)
    private String fileUrl;

    @Column(name = "thumbnail_url", length = 1000)
    private String thumbnailUrl;

    @Column(name = "file_size", nullable = false)
    private Long fileSize;

    @Column(name = "file_type", nullable = false, length = 50)
    private String fileType;

    @Enumerated(EnumType.STRING)
    @Column(name = "file_category", nullable = false, length = 32)
    private FileCategory fileCategory = FileCategory.OTHER;

    @Column(name = "file_storage", nullable = false, length = 20)
    private String fileStorage = "OSS";

    @Column(name = "file_hash", length = 64)
    private String fileHash;

    @Column(name = "business_tag", length = 50)
    private String businessTag;

    @Column(name = "description", length = 500)
    private String description;

    @Column(name = "uploaded_by", nullable = false)
    private Long uploadedBy;

    @Column(name = "uploaded_at", nullable = false)
    private LocalDateTime uploadedAt;

    @Column(name = "upload_source", nullable = false, length = 20)
    private String uploadSource = "WEB";

    /**
     * 关联实体类型 — 白名单约束 (DB CHECK constraint).
     *
     * <p>新增值时必须同步更新:
     * <ol>
     *   <li>本枚举</li>
     *   <li>Flyway migration 的 {@code chk_att_entity_type} CHECK constraint</li>
     * </ol>
     */
    public enum EntityType {
        CUSTOMER,
        CUSTOMER_TRACKING,
        PURCHASE_ORDER,
        PURCHASE_RECEIPT,
        QUALITY_CHECK,
        PRODUCTION_BATCH,
        PAYMENT_VOUCHER,
        INVOICE,
        RD_SAMPLE,
        RECEIPT,
        RETURN_ORDER,
        SHIPMENT,
        WASTAGE_RECORD,
        GROUP_LEADER_REPORT,
        EXPENSE_REPORT,
        LEAVE_REQUEST,
        TIMECLOCK_PHOTO,
        GENERIC
    }

    /**
     * 文件分类. 用于列表 / 过滤 / 图标选择.
     */
    public enum FileCategory {
        PHOTO,
        VIDEO,
        DOCUMENT,
        VOUCHER,
        SIGNATURE,
        OTHER
    }
}

File: web-admin/src/api/attachment.ts
/**
 * 附件 API — C-ATT-1.
 *
 * 对应后端 AttachmentController. 8 endpoint 完整封装.
 * Base: /api/mobile/{factoryId}/attachments (request baseURL='/api/mobile')
 */
import request from './request';

/** 关联实体类型 — 后端 Attachment.EntityType 枚举完整复刻 (18 值). */
export type AttachmentEntityType =
  | 'CUSTOMER'
  | 'CUSTOMER_TRACKING'
  | 'PURCHASE_ORDER'
  | 'PURCHASE_RECEIPT'
  | 'QUALITY_CHECK'
  | 'PRODUCTION_BATCH'
  | 'PAYMENT_VOUCHER'
  | 'INVOICE'
  | 'RD_SAMPLE'
  | 'RECEIPT'
  | 'RETURN_ORDER'
  | 'SHIPMENT'
  | 'WASTAGE_RECORD'
  | 'GROUP_LEADER_REPORT'
  | 'EXPENSE_REPORT'
  | 'LEAVE_REQUEST'
  | 'TIMECLOCK_PHOTO'
  | 'GENERIC';

export type AttachmentFileCategory =
  | 'PHOTO'
  | 'VIDEO'
  | 'DOCUMENT'
  | 'VOUCHER'
  | 'SIGNATURE'
  | 'OTHER';

export interface Attachment {
  id: string;
  factoryId: string;
  entityType: AttachmentEntityType;
  entityId: string;
  fileName: string;
  fileUrl: string;
  thumbnailUrl?: string;
  fileSize: number;
  fileType: string;
  fileCategory: AttachmentFileCategory;
  fileStorage: string;
  fileHash?: string;
  businessTag?: string;
  description?: string;
  uploadedBy: number;
  uploadedAt: string;
  uploadSource: string;
}

export interface RegisterAttachmentRequest {
  entityType: AttachmentEntityType;
  entityId: string;
  fileName: string;
  fileUrl: string;
  thumbnailUrl?: string;
  fileSize: number;
  fileType: string;
  fileCategory?: AttachmentFileCategory;
  fileStorage?: string;
  fileHash?: string;
  businessTag?: string;
  description?: string;
  uploadSource?: string;
}

export interface UpdateAttachmentRequest {
  description?: string;
  businessTag?: string;
  fileCategory?: AttachmentFileCategory;
}

export interface UploadUrlResponse {
  uploadUrl: string;
  fileUrl: string;
}

function getFactoryId(override?: string): string {
  if (override) return override;
  const raw = localStorage.getItem('cretas_user');
  if (!raw || raw === 'undefined' || raw === 'null') {
    throw new Error('未登录 — 请重新登录');
  }
  let parsed: unknown;
  try {
    parsed = JSON.parse(raw);
  } catch {
    throw new Error('登录态损坏 — 请重新登录');
  }
  // User object 结构: { factoryUser: { factoryId, ... }, ... } (auth store schema)
  // Also tolerate top-level factoryId for platform/legacy callers.
  const p = parsed as { factoryId?: string; factoryUser?: { factoryId?: string } } | null;
  const fid = p?.factoryUser?.factoryId ?? p?.factoryId;
  if (!fid) {
    throw new Error('当前账号未绑定工厂 — 无法查看/上传附件');
  }
  return fid;
}

function basePath(factoryId?: string): string {
  return `/${getFactoryId(factoryId)}/attachments`;
}

/** 1. 列表 — 某实体的所有附件. */
export function listAttachments(
  entityType: AttachmentEntityType,
  entityId: string,
  factoryId?: string,
): Promise<{ data: Attachment[]; success: boolean }> {
  return request.get(`${basePath(factoryId)}?entityType=${entityType}&entityId=${encodeURIComponent(entityId)}`);
}

/** 2. 详情. */
export function getAttachment(id: string, factoryId?: string): Promise<{ data: Attachment }> {
  return request.get(`${basePath(factoryId)}/${id}`);
}

/** 3. 下载 — 后端 302 重定向. 直接打开 URL 即可. */
export function downloadAttachmentUrl(id: string, factoryId?: string): string {
  return `/api/mobile${basePath(factoryId)}/${id}/download`;
}

/** 4. 取 OSS pre-signed PUT URL — 5min. */
export function getUploadUrl(
  fileName: string,
  fileType: string,
  factoryId?: string,
): Promise<{ data: UploadUrlResponse }> {
  return request.post(`${basePath(factoryId)}/upload-url`, { fileName, fileType });
}

/** 5. 注册元数据 (OSS 直传完后). */
export function registerAttachment(
  req: RegisterAttachmentRequest,
  factoryId?: string,
): Promise<{ data: Attachment }> {
  return request.post(basePath(factoryId), req);
}

/** 6. 更新 description/businessTag/fileCategory. */
export function updateAttachment(
  id: string,
  req: UpdateAttachmentRequest,
  factoryId?: string,
): Promise<{ data: Attachment }> {
  return request.put(`${basePath(factoryId)}/${id}`, req);
}

/** 7. 软删除. */
export function deleteAttachment(id: string, factoryId?: string): Promise<unknown> {
  return request.delete(`${basePath(factoryId)}/${id}`);
}

/** 8. 批量计数 — 列表页徽章. */
export function batchCountAttachments(
  entityType: AttachmentEntityType,
  entityIds: string[],
  factoryId?: string,
): Promise<{ data: Record<string, number> }> {
  return request.post(`${basePath(factoryId)}/batch-by-entity`, { entityType, entityIds });
}

/**
 * 端到端上传辅助 — 浏览器 File → getUploadUrl → PUT to OSS → register.
 */
export async function uploadAndRegister(
  file: File,
  entityType: AttachmentEntityType,
  entityId: string,
  extras?: { businessTag?: string; description?: string; fileCategory?: AttachmentFileCategory },
  factoryId?: string,
): Promise<Attachment> {
  const urlResp = await getUploadUrl(file.name, file.type, factoryId);
  const { uploadUrl, fileUrl } = urlResp.data;

  const putRes = await fetch(uploadUrl, {
    method: 'PUT',
    headers: { 'Content-Type': file.type },
    body: file,
  });
  if (!putRes.ok) throw new Error(`OSS 上传失败 HTTP ${putRes.status}`);

  const saved = await registerAttachment(
    {
      entityType,
      entityId,
      fileName: file.name,
      fileUrl,
      fileSize: file.size,
      fileType: file.type,
      fileCategory: extras?.fileCategory,
      businessTag: extras?.businessTag,
      description: extras?.description,
      uploadSource: 'WEB',
    },
    factoryId,
  );
  return saved.data;
}
